runtime/near-vm-logic/src/logic.rs

```diff
@@ -278,49 +285,61 @@ impl<'a> VMLogic<'a> {
     fn get_utf8_string(&mut self, len: u64, ptr: u64) -> Result<String> {
         let mut buf;
+        let max_len = self.config.max_log_len;
         if len != std::u64::MAX {
-            if len > self.config.max_log_len {
+            if len > max_len {
                 return Err(HostError::BadUTF8.into());
             }
             buf = Self::memory_get(self.memory, ptr, len)?;
         } else {
             buf = vec![];
-            for i in 0..=self.config.max_log_len {
-                Self::try_fit_mem(self.memory, ptr, i)?;
+            for i in 0..=max_len {
+                if i == max_len {
+                    return Err(HostError::BadUTF8.into());
+                }
+                Self::try_fit_mem(self.memory, ptr + i, 1)?;
                 let el = self.memory.read_memory_u8(ptr + i);
                 if el == 0 {
                     break;
                 }
-                if i == self.config.max_log_len {
-                    return Err(HostError::BadUTF8.into());
-                }
                 buf.push(el);
             }
         }
         String::from_utf8(buf).map_err(|_| HostError::BadUTF8.into())
     }
 
-    /// Helper function to read UTF-16 assembly script formatted string from guest memory.
+    /// Helper function to read UTF-16 formatted string from guest memory.
     /// # Errors
     ///
     /// * If string extends outside the memory of the guest with `MemoryAccessViolation`;
     /// * If string is not UTF-16 returns `BadUtf16`.
-    fn get_assembly_script_utf16_string(&mut self, ptr: u64) -> Result<String> {
-        if ptr < 4 {
-            return Err(HostError::BadUTF16.into());
-        }
-        let mut slice = [0u8; 4];
-        let buf = Self::memory_get(self.memory, ptr - 4, 4)?;
-        slice.copy_from_slice(&buf);
-        let len: u32 = u32::from_le_bytes(slice);
-        if len % 2 != 0 {
-            return Err(HostError::BadUTF16.into());
-        }
-        let buffer = Self::memory_get(self.memory, ptr, u64::from(len))?;
+    fn get_utf16_string(&mut self, len: u64, ptr: u64) -> Result<String> {
         let mut u16_buffer = Vec::new();
-        for i in 0..((len / 2) as usize) {
-            let c = u16::from(buffer[i * 2]) + u16::from(buffer[i * 2 + 1]) * 0x100;
-            u16_buffer.push(c);
+        let max_len = self.config.max_log_len;
+        if len != std::u64::MAX {
+            let input = Self::memory_get(self.memory, ptr, len as u64)?;
+            if len % 2 != 0 || len > max_len {
+                return Err(HostError::BadUTF16.into());
+            }
+            for i in 0..((len / 2) as usize) {
+                u16_buffer.push(u16::from_le_bytes([input[i as usize * 2], input[i as usize * 2 + 1]]));
+            }
+        } else {
+            // Takes 2 bytes each iter
+            for i in 0..=(max_len / size_of::<u16>() as u64) {
+                if i * size_of::<u16>() as u64 == max_len {
```

Comment:
If `max_len` is not divisible by 2, this would fail, so the method will return a string without null-termination.

Reply:
adressed